src/bin/shard/chat.rs

```diff
@@ -218,6 +218,108 @@ pub fn send_group_menuchat_message(
     Ok(())
 }
 
+pub fn send_buddy_freechat_message(
+    clients: &mut ClientMap,
+    state: &mut ShardServerState,
+) -> FFResult<()> {
+    let client = clients.get_self();
+    let pkt: sP_CL2FE_REQ_SEND_BUDDY_FREECHAT_MESSAGE =
+        *client.get_packet(P_CL2FE_REQ_SEND_BUDDY_FREECHAT_MESSAGE)?;
+    let pc_id = client.get_player_id()?;
+    let player = state.get_player(pc_id)?;
+
+    let buddy_uid = pkt.iBuddyPCUID;
+
+    let msg = util::parse_utf16(&pkt.szFreeChat)?;
+    if !helpers::validate_menuchat_message(&msg) {
+        return Err(FFError::build(
+            Severity::Warning,
+            format!("Invalid menuchat message\n\t{}: '{}'", player, msg),
+        ));
+    }
+
+    let response_pkt = sP_FE2CL_REP_SEND_BUDDY_FREECHAT_MESSAGE_SUCC {
+        iFromPCUID: player.get_uid(),
+        iToPCUID: pkt.iBuddyPCUID,
+        szFreeChat: pkt.szFreeChat,
+        iEmoteCode: pkt.iEmoteCode,
+    };
+
+    if let Some(buddy) = state.get_player_by_uid(buddy_uid) {
+        if let Some(buddy_client) = buddy.get_client(clients) {
+            buddy_client
+                .send_packet(P_FE2CL_REP_SEND_BUDDY_FREECHAT_MESSAGE_SUCC, &response_pkt)?;
+            let sender = state.get_player_by_uid(response_pkt.iFromPCUID).unwrap();
+            let sender_client = sender.get_client(clients).unwrap();
+            sender_client
+                .send_packet(P_FE2CL_REP_SEND_BUDDY_FREECHAT_MESSAGE_SUCC, &response_pkt)?;
+            return Ok(());
+        }
+    }
+
+    let login_server = clients.get_login_server().unwrap();
```

Comment:
Using `unwrap` here is very dangerous, because if communication to the login server is lost, this will crash the shard server. We should log a warning and stop if the login server isn't connected.
```suggestion
    if let Some(login_server) = clients.get_login_server() {
        ...
    } else {
        // warn here
    }
```